go/vt/vttablet/onlineddl/executor.go

```diff
@@ -1696,6 +1703,80 @@ func (e *Executor) CancelPendingMigrations(ctx context.Context, message string)
 	return result, nil
 }
 
+func (e *Executor) validateThrottleParams(ctx context.Context, expireString string, ratioLiteral *sqlparser.Literal) (duration time.Duration, ratio float64, err error) {
+	duration = time.Hour * 24 * 365 * 100
+	if expireString != "" {
+		duration, err = time.ParseDuration(expireString)
+		if err != nil || duration < 0 {
+			return duration, ratio, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid EXPIRE value: %s. Try '120s', '30m', '1h', etc. Allowed units are (s)ec, (m)in, (h)hour", expireString)
+		}
+	}
+	ratio = 1.0
+	if ratioLiteral != nil {
+		ratio, err = strconv.ParseFloat(ratioLiteral.Val, 64)
+		if err != nil || ratio < 0 || ratio > 1 {
+			return duration, ratio, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid RATIO value: %s. Try any decimal number between '0.0' (no throttle) and `1.0` (fully throttled)", ratioLiteral.Val)
+		}
+	}
+	return duration, ratio, nil
+}
+
+// ThrottleMigration
+func (e *Executor) ThrottleMigration(ctx context.Context, uuid string, expireString string, ratioLiteral *sqlparser.Literal) (result *sqltypes.Result, err error) {
+	duration, ratio, err := e.validateThrottleParams(ctx, expireString, ratioLiteral)
+	if err != nil {
+		return nil, err
+	}
+	if !e.lagThrottler.IsEnabled() {
+		return nil, throttle.ErrThrottlerNotEnabled
+	}
+	if !e.lagThrottler.IsOpen() {
+		return nil, throttle.ErrThrottlerNotEnabled
+	}
+	_ = e.lagThrottler.ThrottleApp(uuid, time.Now().Add(duration), ratio)
+	return emptyResult, nil
+}
+
+// ThrottleAllMigrations
+func (e *Executor) ThrottleAllMigrations(ctx context.Context, expireString string, ratioLiteral *sqlparser.Literal) (result *sqltypes.Result, err error) {
+	duration, ratio, err := e.validateThrottleParams(ctx, expireString, ratioLiteral)
+	if err != nil {
+		return nil, err
+	}
+	if !e.lagThrottler.IsEnabled() {
+		return nil, throttle.ErrThrottlerNotEnabled
+	}
+	if !e.lagThrottler.IsOpen() {
+		return nil, throttle.ErrThrottlerNotEnabled
+	}
+	_ = e.lagThrottler.ThrottleApp(throttlerOnlineDDLApp, time.Now().Add(duration), ratio)
+	return emptyResult, nil
+}
+
+// UnthrottleMigration
+func (e *Executor) UnthrottleMigration(ctx context.Context, uuid string) (result *sqltypes.Result, err error) {
+	if !e.lagThrottler.IsEnabled() {
+		return nil, throttle.ErrThrottlerNotEnabled
+	}
+	if !e.lagThrottler.IsOpen() {
+		return nil, throttle.ErrThrottlerNotEnabled
+	}
+	_ = e.lagThrottler.UnthrottleApp(uuid)
+	return emptyResult, nil
+}
+
+// UnthrottleAllMigrations
+func (e *Executor) UnthrottleAllMigrations(ctx context.Context) (result *sqltypes.Result, err error) {
+	if !e.lagThrottler.IsEnabled() {
```

Comment:
nit: you are doing the same checks in three places here. maybe `lagThrottler` could have an `IsReady() error` that returns the correct error if the throttler is not ready?

Reply:
fixed